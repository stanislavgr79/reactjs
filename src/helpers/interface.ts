export interface IMovie {
  [x: string]: any;
  id?: number;
  title: string;
  tagline?: string;
  vote_average?: number;
  vote_count?: number;
  release_date: string;
  movieUrl: string;
  genres: string[] | { label: string, value: string }[];
  poster_path?: string | undefined;
  overview: string;
  runtime: number | string | undefined;
}
