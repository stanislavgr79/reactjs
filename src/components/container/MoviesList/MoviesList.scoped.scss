.movies_finded {
    font-size: 20px;
    color: #BBB;
}

.movies_not_finded {
    color: #BBB;
    font-size: 60px;
    text-align: center;
    padding-top: 15%;
}