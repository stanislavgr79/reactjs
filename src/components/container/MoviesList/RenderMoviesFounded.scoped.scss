.movies-wraper {
    padding-bottom: 50px;
    padding-top: 10px;
    .movie-unit {
        position: relative;
        .movie_holder {
            position: relative;
        }
        .sub_menu_icon {
            position: absolute !important;
            right: 7px;
            top: 7px;
            width: 50px;
            height: 50px;
            &:hover {
                cursor: pointer;
            }
        }
    }
}