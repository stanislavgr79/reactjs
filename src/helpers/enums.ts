/* eslint-disable no-unused-vars */

/**
 * Enum for common SelectGenres.
 * @readonly
 * @enum {{label: string, value: string}}
 */
export const selectGenres = Object.freeze([
  { label: 'Action', value: 'Action' },
  { label: 'Adventure', value: 'Adventure' },
  { label: 'Animation', value: 'Animation' },
  { label: 'Comedy', value: 'Comedy' },
  { label: 'Documentary', value: 'Documentary' },
  { label: 'Drama', value: 'Drama' },
  { label: 'Family', value: 'Family' },
  { label: 'Horror', value: 'Horror' },
  { label: 'Mystery', value: 'Mystery' },
  { label: 'Thriller', value: 'Thriller' },
]);

/**
 * Enum for common Genres.
 * @readonly
 * @enum { string }
 */
export enum Genres {
  ALL = '',
  DOCUMENTARY = 'documentary',
  COMEDY = 'comedy',
  HORROR = 'horror',
  CRIME = 'crime',
}

/**
 * Enum for common SortBy.
 * @readonly
 * @enum { string }
 */
export enum SortBy {
  TITLE = 'title',
  RELEASE = 'release_date',
  RUNTIME = 'runtime',
  RATING = 'vote_average',
}

/**
 * Enum for common SortOrder.
 * @readonly
 * @enum { string }
 */
export enum SortOrder {
  ASC = 'asc',
  DESC = 'desc',
}
