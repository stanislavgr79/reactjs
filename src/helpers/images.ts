const images = [
  {
    name: 'Ricky and Morty.jpg',
    value: './resources/images/Ricky and Morty.jpg',
  },
  {
    name: 'Pulp Fiction.jpg',
    value: './resources/images/Pulp Fiction.jpg',
  },
  {
    name: 'Kill Bill.jpg',
    value: './resources/images/Kill Bill.jpg',
  },
  {
    name: 'Avengers.jpg',
    value: './resources/images/Avengers.jpg',
  },
  {
    name: 'Inception.jpg',
    value: './resources/images/Inception.jpg',
  },
  {
    name: 'Reservoir Dogs.jpg',
    value: './resources/images/Reservoir Dogs.jpg',
  },
];

export default images;
