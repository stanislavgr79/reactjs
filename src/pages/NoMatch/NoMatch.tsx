import React from 'react';

import Header from '../../components/structure/Header';
import Footer from '../../components/structure/Footer';

import './NoMatch.scoped.scss';
import { useLocation } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { AppState } from '../../redux/redux-store';

export default function NoMatch(): JSX.Element {
  const unknownPath = useLocation();
  const { location } = useSelector((store: AppState) => {
    return store.searchStore;
  });
  return (
    <div className="nomatch-page">
      <Header />
      <div className="nomatch-page-panel">
        <div id="nomatch-header">
          <div id="title-heading" className="pagetitle">
            <h1 id="title-text">Page Not Found</h1>
          </div>
        </div>
        <div id="sidebar-container"></div>
        <p>We can&apos;t find that page. This could be because:</p>
        <ul>
          <li>The page {location != '' ? location : unknownPath.pathname} doesn&apos;t exist.</li>
          <li>The page exists, but you don&apos;t have view permission for that space.</li>
        </ul>
        <div id="action-messages">
          <h3>Go to</h3>
          <ul>
            <li>
              <a href="/movies">Site Homepage</a>
            </li>
          </ul>
        </div>
      </div>
      <Footer />
    </div>
  );
}
