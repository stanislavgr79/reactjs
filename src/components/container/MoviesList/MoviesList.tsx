import React, { useEffect, useMemo, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppState } from '../../../redux/redux-store';
import { updateSelectedMovieId } from '../../../redux/actions/movies-actions';

import MovieCard from '../../atom/MovieCard';
import MovieMenuPopup from '../MovieMenuPopup';
import { IMovie } from '../../../helpers/interface';

import image from '@resources/images/menu-icon.png';
import './MoviesList.scoped.scss';

interface IProp {
  movies: IMovie[];
}

const renderMoviesNotFounded = () => {
  <div className="movies_not_finded">Movies not finded</div>;
};

const RenderMoviesFounded = ({ movies }: IProp): JSX.Element => {
  const [showSubMenu, setShowSubMenu] = useState(false);
  const [activeMovieId, setActiveMovieId] = useState(0);
  const [selectedMovieId, setSelectedMovieId] = useState(0);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(updateSelectedMovieId(selectedMovieId));
  }, [dispatch, selectedMovieId]);

  const updateShowSubMenuWithId = (id: number) => {
    setActiveMovieId(id);
    setShowSubMenu(!showSubMenu);
  };

  const updateShowSubMenu = () => {
    setShowSubMenu(!showSubMenu);
  };

  const changeSelectedMovieId = (id: number) => {
    setSelectedMovieId(id);
  };

  return (
    <>
      <div className="container-md movies-wraper">
        <div className="row row-cols-3">
          {movies.map((movie: IMovie) => (
            <div className="col movie-unit" key={movie.title}>
              <div className="movie_holder">
                <MovieCard
                  id={movie.id}
                  title={movie.title}
                  releaseDate={movie.release_date}
                  genres={movie.genres}
                  movieUrl={movie.movieUrl}
                  posterPath={movie.poster_path}
                  key={movie.title}
                  changeSelectedMovieId={changeSelectedMovieId}
                />
                {!showSubMenu ? (
                  <img
                    src={image}
                    className="sub_menu_icon"
                    onClick={() => updateShowSubMenuWithId(movie.id)}
                  />
                ) : activeMovieId == movie.id ? (
                  <MovieMenuPopup movie={movie} closeMenu={updateShowSubMenu} />
                ) : (
                  <img src={image} className="sub_menu_icon" onClick={updateShowSubMenu} />
                )}
              </div>
            </div>
          ))}
        </div>
      </div>
    </>
  );
};

function filterByGenre(movies: IMovie[], genre: string) {
  return movies.filter(function (el) {
    let isFinded = false;
    el.genres.forEach(function (e) {
      if (e.toLowerCase() == genre.toLowerCase()) {
        isFinded = true;
      }
    });
    if (isFinded) {
      return el;
    }
  });
}

function dynamicSort(property: string) {
  let sortOrder = 1;
  if (property.toString().startsWith('-')) {
    sortOrder = -1;
    property = property.toString().substr(1);
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return function (a: { [x: string]: any }, b: { [x: string]: any }) {
    if (property == 'release') {
      property = 'release_date';
    }

    const result = a[property] < b[property] ? -1 : a[property] > b[property] ? 1 : 0;
    return result * sortOrder;
  };
}

function searchMovies(movies: IMovie[], search: string) {
  return movies.filter((movie: IMovie) => movie.title.toLowerCase().includes(search.toLowerCase()));
}

export default function MoviesList(): JSX.Element {
  const store = useSelector((store: AppState) => {
    return {
      movies: store.moviesStore.movies,
      sidebar: store.sidebar,
      search: store.searchStore.search,
    };
  });

  const movies: IMovie[] = useMemo(() => {
    const movies: IMovie[] =
      store.search.value != '' ? searchMovies(store.movies, store.search.value) : store.movies;
    const filteredMovies: IMovie[] =
      store.sidebar.genre == 'ALL' ? movies : filterByGenre(movies, store.sidebar.genre);
    return filteredMovies.sort(dynamicSort(store.sidebar.sortBy.toString().toLowerCase()));
  }, [store.movies, store.search, store.sidebar.genre, store.sidebar.sortBy]);

  return (
    <>
      <div className="container-md movies_finded">
        <label>
          <strong>{movies.length}</strong> movies found
        </label>
      </div>
      {movies.length > 0 ? <RenderMoviesFounded movies={movies} /> : renderMoviesNotFounded}
    </>
  );
}
