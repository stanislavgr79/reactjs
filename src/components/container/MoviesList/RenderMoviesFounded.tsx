import React, { useState } from 'react';

import MovieCard from '../../atom/MovieCard';
import MovieMenuPopup from '../MovieMenuPopup';
import { IMovie } from '../../../helpers/interface';

import image from '../../../resources/images/menu-icon.png';
import './RenderMoviesFounded.scoped.scss';

interface IProp {
  movies: IMovie[];
}

export default function RenderMoviesFounded({ movies }: IProp): JSX.Element {
  const [showSubMenu, setShowSubMenu] = useState(false);
  const [activeMovieId, setActiveMovieId] = useState(0);

  const updateShowSubMenuWithId = (id: number) => {
    setActiveMovieId(id);
    setShowSubMenu(!showSubMenu);
  };

  const updateShowSubMenu = () => {
    setShowSubMenu(!showSubMenu);
  };

  return (
    <>
      <div className="container-md movies-wraper">
        <div className="row row-cols-3">
          {movies.map((movie: IMovie, index: number) => (
            <div className="col movie-unit" key={`unit-${index}`}>
              <div className="movie_holder">
                <MovieCard movie={movie} />
                {!showSubMenu ? (
                  <img
                    src={image}
                    className="sub_menu_icon first"
                    onClick={() => updateShowSubMenuWithId(Number(movie.id))}
                  />
                ) : activeMovieId == movie.id ? (
                  <MovieMenuPopup movie={movie} closeMenu={updateShowSubMenu} />
                ) : (
                  <img src={image} className="sub_menu_icon second" onClick={updateShowSubMenu} />
                )}
              </div>
            </div>
          ))}
        </div>
      </div>
    </>
  );
}
