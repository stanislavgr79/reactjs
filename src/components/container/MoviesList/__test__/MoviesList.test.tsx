import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import { render } from '@testing-library/react';

import MoviesList from '../MoviesList';
import { Genres, SortBy, SortOrder } from '../../../../helpers/enums';

import { apiMiddleware } from 'redux-api-middleware';
import fetchMock from 'fetch-mock';
import thunk from 'redux-thunk';
import NoMatch from '../../../../pages/NoMatch';
const middlewares = [thunk, apiMiddleware];
const URL_API_MOVIES = 'HTTP://localhost:4000/movies';

jest.mock('../../../../pages/NewPage', () => {
  return {
    __esModule: true,
    default: () => {
      return <div className="NewPage"></div>;
    },
  };
});
jest.mock('../RenderMoviesFounded', () => {
  return {
    __esModule: true,
    default: () => {
      return <div className="RenderMoviesFounded"></div>;
    },
  };
});
jest.mock('../../../../pages/NoMatch', () => {
  return {
    __esModule: true,
    default: () => {
      return <div className="NoMatch"></div>;
    },
  };
});
const parameters = [
  ['offset', '0'],
  ['limit', '6'],
  ['searchBy', 'title'],
  ['search', ''],
  ['filter', ''],
  ['sortBy', 'title'],
  ['sortOrder', 'asc'],
];
fetchMock.mock(`${URL_API_MOVIES}?${new URLSearchParams(parameters).toString()}`, 200);

describe('MoviesList', () => {
  const movie = {
    id: 12345,
    title: 'film title',
    release_date: '2020-01-02',
    poster_path: null,
    genres: ['crime'],
    overview: 'overview',
    runtime: 66,
    vote_average: 5.5,
    tagline: 0,
  };
  const mockStore = configureStore(middlewares);

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('render newPage', () => {
    const initialState = {
      moviesStore: {
        movies: {
          totalAmount: 1,
          data: [movie],
          offset: 0,
          limit: 6,
        },
        newPage: true,
        dataStatus: 'idle',
      },
      sidebar: { genre: Genres.ALL, sortBy: SortBy.TITLE, sortOrder: SortOrder.ASC },
      searchStore: { search: { value: '' } },
    };

    const store = mockStore(initialState);

    const tree = render(
      <BrowserRouter>
        <Provider store={store}>
          <MoviesList />
        </Provider>
      </BrowserRouter>,
    );

    expect(tree.baseElement).toMatchSnapshot();
  });

  test('render redirect to 404 ', () => {
    const initialState = {
      moviesStore: {
        movies: {
          totalAmount: 1,
          data: [movie],
          offset: 0,
          limit: 6,
        },
        newPage: true,
        dataStatus: 'failed',
      },
      sidebar: { genre: Genres.ALL, sortBy: SortBy.TITLE, sortOrder: SortOrder.ASC },
      searchStore: { search: { value: '' } },
    };

    const store = mockStore(initialState);

    const tree = render(
      <BrowserRouter>
        <Switch>
          <Route exact key="404" path="/404" render={() => <NoMatch />} />
        </Switch>
        <Provider store={store}>
          <MoviesList />
        </Provider>
      </BrowserRouter>,
    );

    expect(tree.baseElement).toMatchSnapshot();
  });

  test('render data Loading', () => {
    const initialState = {
      moviesStore: {
        movies: {
          totalAmount: 1,
          data: [movie],
          offset: 0,
          limit: 6,
        },
        newPage: true,
        dataStatus: 'loading',
      },
      sidebar: { genre: Genres.ALL, sortBy: SortBy.TITLE, sortOrder: SortOrder.ASC },
      searchStore: { search: { value: '' } },
    };

    const store = mockStore(initialState);

    const tree = render(
      <BrowserRouter>
        <Provider store={store}>
          <MoviesList />
        </Provider>
      </BrowserRouter>,
    );

    expect(tree.baseElement).toMatchSnapshot();
  });

  test('render movies not founded', () => {
    const initialState = {
      moviesStore: {
        movies: {
          totalAmount: 0,
          data: [],
          offset: 0,
          limit: 6,
        },
        newPage: false,
        dataStatus: 'success',
      },
      sidebar: { genre: Genres.ALL, sortBy: SortBy.TITLE, sortOrder: SortOrder.ASC },
      searchStore: { search: { value: '' } },
    };

    const store = mockStore(initialState);

    const tree = render(
      <BrowserRouter>
        <Provider store={store}>
          <MoviesList />
        </Provider>
      </BrowserRouter>,
    );

    expect(tree.baseElement).toMatchSnapshot();
  });

  test('render movies not founded', () => {
    const initialState = {
      moviesStore: {
        movies: {
          totalAmount: 0,
          data: [],
          offset: 0,
          limit: 6,
        },
        newPage: false,
        dataStatus: 'success',
      },
      sidebar: { genre: Genres.ALL, sortBy: SortBy.TITLE, sortOrder: SortOrder.ASC },
      searchStore: { search: { value: '' } },
    };

    const store = mockStore(initialState);

    const tree = render(
      <BrowserRouter>
        <Provider store={store}>
          <MoviesList />
        </Provider>
      </BrowserRouter>,
    );

    expect(tree.baseElement).toMatchSnapshot();
  });

  test('render movies founded', () => {
    const initialState = {
      moviesStore: {
        movies: {
          totalAmount: 1,
          data: [movie],
          offset: 0,
          limit: 6,
        },
        newPage: false,
        dataStatus: 'success',
      },
      sidebar: { genre: Genres.ALL, sortBy: SortBy.TITLE, sortOrder: SortOrder.ASC },
      searchStore: { search: { value: '' } },
    };

    const store = mockStore(initialState);

    const tree = render(
      <BrowserRouter>
        <Provider store={store}>
          <MoviesList />
        </Provider>
      </BrowserRouter>,
    );

    expect(tree.baseElement).toMatchSnapshot();
  });
});
