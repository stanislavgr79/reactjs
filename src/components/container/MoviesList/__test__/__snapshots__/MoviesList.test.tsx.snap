// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesList render data Loading 1`] = `
<body>
  <div>
    <div
      class="loader"
    >
      Loading...
    </div>
  </div>
</body>
`;

exports[`MoviesList render movies founded 1`] = `
<body>
  <div>
    <div
      class="container-md movies_finded"
    >
      <label>
        <strong>
          1
        </strong>
         movies found
      </label>
    </div>
    <div
      class="RenderMoviesFounded"
    />
  </div>
</body>
`;

exports[`MoviesList render movies not founded 1`] = `
<body>
  <div>
    <div
      class="container-md movies_finded"
    >
      <label>
        <strong>
          0
        </strong>
         movies found
      </label>
    </div>
    <div
      class="movies_not_finded"
    >
      Movies not finded
    </div>
  </div>
</body>
`;

exports[`MoviesList render movies not founded 2`] = `
<body>
  <div>
    <div
      class="container-md movies_finded"
    >
      <label>
        <strong>
          0
        </strong>
         movies found
      </label>
    </div>
    <div
      class="movies_not_finded"
    >
      Movies not finded
    </div>
  </div>
</body>
`;

exports[`MoviesList render newPage 1`] = `
<body>
  <div>
    <div
      class="new-page"
    >
      <div
        class="new-page-container"
      >
        new page
        /
      </div>
      ;
    </div>
  </div>
</body>
`;

exports[`MoviesList render redirect to 404  1`] = `
<body>
  <div>
    <div
      class="NoMatch"
    />
  </div>
</body>
`;
